sdk/helper/aliasmetadata/alias_metadata.go

```diff
@@ -0,0 +1,212 @@
+		return fmt.Errorf("%q contains an unavailable field, please select from %q",
+			strings.Join(userProvided, ", "), strings.Join(h.fields.all(), ", "))
+	}
+	h.aliasMetadata = userProvided
+	return nil
+}
+
+// expandDefaultField looks for the field "default" and, if exists,
+// replaces it with the actual default fields signified.
+func (h *Handler) expandDefaultField(userProvided []string) (expanded []string) {
+	for _, field := range userProvided {
+		if field != "default" {
+			expanded = append(expanded, field)
+			continue
+		}
+		for _, dfltField := range h.fields.Default {
+			expanded = append(expanded, dfltField)
+		}
+	}
+	return expanded
+}
+
+// PopulateDesiredAliasMetadata is intended to be used during login
+// just before returning an auth.
+// It takes the available alias metadata and,
+// if the auth should have it, adds it to the auth's alias metadata.
+func (h *Handler) PopulateDesiredAliasMetadata(auth *logical.Auth, available map[string]string) error {
+	if auth == nil {
+		return errors.New("auth is nil")
+	}
+	if auth.Alias == nil {
+		return errors.New("auth alias is nil")
+	}
+	if auth.Alias.Name == "" {
+		// We need the caller to set the alias name or there will
+		// be nothing for these fields to operate upon.
+		return errors.New("auth alias name must be set")
+	}
+	if auth.Alias.Metadata == nil {
+		auth.Alias.Metadata = make(map[string]string)
+	}
+	fieldsToInclude := h.fields.Default
+	if h.aliasMetadata != nil {
+		fieldsToInclude = h.aliasMetadata
+	}
+	for availableField, itsValue := range available {
+		if strutil.StrListContains(fieldsToInclude, availableField) {
+			auth.Alias.Metadata[availableField] = itsValue
+		}
+	}
+	return nil
+}
+
+func (h *Handler) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&struct {
+		AliasMetadata []string `json:"alias_metadata"`
+	}{
+		AliasMetadata: h.aliasMetadata,
+	})
+}
+
+func (h *Handler) UnmarshalJSON(data []byte) error {
+	jsonable := &struct {
+		AliasMetadata []string `json:"alias_metadata"`
+	}{
+		AliasMetadata: h.aliasMetadata,
+	}
+	if err := json.Unmarshal(data, jsonable); err != nil {
+		return err
+	}
+	h.aliasMetadata = jsonable.AliasMetadata
+	return nil
+}
+
+func description(fields *Fields) string {
+	desc := "The metadata to include on the aliases generated by this plugin."
+	if len(fields.Default) > 0 {
+		desc += fmt.Sprintf(" When set to 'default', includes: %s.", strings.Join(fields.Default, ", "))
+	}
+	if len(fields.AvailableToAdd) > 0 {
+		desc += fmt.Sprintf(" These fields are available to add: %s.", strings.Join(fields.AvailableToAdd, ", "))
+	}
+	desc += " Not editing this field means the 'default' fields are included." +
+		" Explicitly setting this field to empty overrides the 'default' and means no alias metadata will be included." +
+		" Add fields by sending, 'default,field1,field2'." +
+		" We advise only including fields that change rarely because each change triggers a storage write."
```

Comment:
I suggest removing this from here as it is highly dependent on the user's situation. e.g. if login frequency is not that high, there is likely no issue. We can elaborate in docs on this, as well as providing specific guidance for those backends that support configurable metadata.

Reply:
This is removed in https://github.com/hashicorp/vault/pull/8783/commits/ae3ba7f04ddcac3eb4f03036e4a6c2ce4028081e